resources/sdk/purecloudjava/templates/ApiClient.mustache

```diff
@@ -975,4 +1055,79 @@ public class ApiClient implements AutoCloseable {
         @Override
         public void close() throws Exception { }
     }
+
+    public static class RetryConfiguration {
+        private long backoffInterval;
+        private int defaultDelay;
+        private int maxRetryTime;
+
+        public RetryConfiguration(RetryConfigBuilder builder) {
+            this.backoffInterval = builder.backoffInterval;
+            this.defaultDelay = builder.defaultDelay;
+            this.maxRetryTime = builder.maxRetryTime;
+        }
+    }
+
+    public static class Retry {
+        private long backoffInterval;
+        private int defaultDelay;
+        private int maxRetryTime;
+        private final static int MAX_RETRIES_BEFORE_BACKOFF = 5;
+        private int retryCountBeforeBackOff = 0;
+        private long waitTimeInMillis = defaultDelay * 1000L;
+        private Stopwatch stopwatch = null;
+
+        private final List<Integer> statusCodes = Arrays.asList(429, 502, 503, 504);
+        private static Logger LOGGER = LoggerFactory.getLogger(Retry.class);
+
+        public Retry(RetryConfiguration retryConfiguration) {
+            this.backoffInterval = retryConfiguration.backoffInterval;
+            this.defaultDelay = retryConfiguration.defaultDelay;
+            this.maxRetryTime = retryConfiguration.maxRetryTime;
+            stopwatch = Stopwatch.createStarted();
+        }
+
+        public boolean shouldRetry(ApiClientConnectorResponse connectorResponse) {
+            return stopwatch.elapsed(TimeUnit.MILLISECONDS) < maxRetryTime * 1000L && shouldWait(connectorResponse);
+        }
+
+        private boolean shouldWait(ApiClientConnectorResponse connectorResponse) {
+            if (statusCodes.contains(connectorResponse.getStatusCode())) {
+
+                if (connectorResponse.getHeaders().containsKey("Retry-After")) {
+                    waitTimeInMillis = Integer.parseInt(connectorResponse.getHeaders().getOrDefault("Retry-After", String.valueOf(defaultDelay))) * 1000L;
+                }
+                //If status code is 429 then wait until retry-after time and retry. OR If status code is retryable then for the first 5 times: wait until retry-after time and retry.
+                if (connectorResponse.getStatusCode() == 429 || retryCountBeforeBackOff++ < MAX_RETRIES_BEFORE_BACKOFF) {
+                    return waitBeforeRetry(waitTimeInMillis);
+                }
+                //If status code is 50x and if 3 Sec and 9 Sec buckets were already retried then wait for every 27 Sec and retry.
+                if (stopwatch.elapsed(TimeUnit.MILLISECONDS) > (backoffInterval * 2)) {
```

Comment:
These three conditions can be replaced with one line: `return waitBeforeRetry(getWaitTimeExp(Math.min(3, Math.floor(stopwatch.elapsed(TimeUnit.MILLISECONDS)/backoffInterval) + 1)))`

Reply:
perfect. thank you.